networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
It looks like we do need to explicitly cast. While you can freely add int and float arrays with NumPy, you cannot do an inplace addition of float data to an int array (in the new casting rules).

For the solution here, I think it makes more sense to make the Laplacian be a float matrix (much earlier in the two user-facing functions) instead of doing it when we create the diagonal matrix. Thoughts on that?